ocaml fails to build on hydra aarch64-darwin

### Describe the bug

ocaml build is failing on hydra but curiously succeeds when building locally... previous versions (ocaml_5_1) are not cached so assume fail to build on hydra but not sure how to access hydra page.

https://hydra.nixos.org/job/nixpkgs/trunk/ocaml.aarch64-darwin
https://hydra.nixos.org/build/275103717

```
[...]
File "_none_", line 1:
Error: I/O error: dllunixbyt.so_The boot compiler cannot load DLLs: No such file or directory
make[3]: *** [Makefile:558: debugger/ocamldebug] Error 2
make[3]: *** Waiting for unfinished jobs....
File "_none_", line 1:
Error: I/O error: dllunixbyt.so_The boot compiler cannot load DLLs: No such file or directory
make[3]: *** [Makefile:544: ocamldoc/ocamldoc] Error 2
```

### Notify maintainers


<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->
@NixOS/ocaml
### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-darwin"`
 - host os: `Darwin 24.0.0, macOS 15.0.1`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.24.9`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

```
ProductName:            macOS
ProductVersion:         15.0.1
BuildVersion:           24A348
```
---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/compilers/ocaml/generic.nix
+++ b/pkgs/development/compilers/ocaml/generic.nix
@@ -105,7 +105,9 @@ stdenv.mkDerivation (args // {
   # sequential order among them as a single rule.
   makefile = ./Makefile.nixpkgs;
   buildFlags = if useNativeCompilers
-    then ["nixpkgs_world_bootstrap_world_opt"]
+    then [(if lib.versionOlder version "5.2"
+      then "nixpkgs_world_bootstrap_world_opt"
+      else "defaultentry")]
     else ["nixpkgs_world"];
   buildInputs = optional (lib.versionOlder version "4.07") ncurses
     ++ optionals useX11 [ libX11 xorgproto ];

